### DISCUSSION TITLE
Import, originals, and metadata extraction

### DISCUSSION THREAD

**Discussion by jvallery** [photoprism/photoprism#691] [Ideas]
I'm looking to leverage Photoprism to manage a large (several TiB) personal image and video library that has many years of neglect and lack of curation. There are many "fuzzy" duplicates and accidentally imported renditions. I've tried to dedupe everything leveraging a binary hash match (jdupes/fdupes with MD5) and have manually cleaned things up as much as possible.  All of this occurred before pointing the Photoprism importer at my library folders. 

My initial observations are that it has done a pretty good job of cleaning things up and further deduplicating. I'm still left with many dupes in the form of renditions (downsized thumbnail versions) with distinct entries in the DB. 

I'd like to understand what "dedupe" actually means for Photoprism.  

The tool I'd like on import would:

while (photos to import) {

- Calculate a "fuzzy" match signature ([various techniques for doing this](http://www.jcomputers.us/vol8/jcp0811-06.pdf)) for the image and compare to existing signatures in the database

if (match in DB) {

- Determine which image (one currently in originals or one being examined) has a higher resolution and replace the image in the "originals" folder
- Compare both images and if the lower resolution one has metadata the higher resolution one does not have extract it and copy it to the higher resolution image
- If a change to metadata results in a new "date taken" update the filename and move to the correct folder in originals library
- Write all metadata (including any tags from Photoprism) back to image file EXIF
- Discard lower resolution file into "Archive" folder
- Update DB and repeat

} else {

- Extract as much metadata as possible from the file (looking for the oldest available date in EXIF to determine date taken as example)
- Copy file to originals based on date taken

}
}
How close to this is Photoprism?

My top feature requests in order of importance:

1) Files moved on the filesystem to an "Archive" folder (which is separate from the originals folder) when archived
2) Ability to fully purge "archive images from the DB 
3) Improvements to the UI and responsiveness for "bulk archive" use cases (shift->click gets painful)
4) Bulk tagging "scans" based on the camera (e.g. CanoScan 9000F) 
5) Hot key support in full screen for star/private/archive/next photo to do quick image reviews
6) GPU support for ffmpeg and tensorflow
7) Facial recognition
   - Tag person in face DB with a "birth date"
   - Use age detection model (HTTP://how-old.net) to approximate date of photo-based on the age of person and birth date
   - Apply approximate year to photos as date taken to allow for chronology of scanned photos

**Comment by dscheffy** [comment]
From what I've seen, so far photoprism only dedupes on exact file content match (as in the sha1 hash of the file has already been seen and indexed, so if a second copy of the file is seen in a different location, it's skipped -- I'm not sure if any of the secondary folder name details update the metadata or not). 

Like you, I was hoping for something a little more sophisticated:

1. Once upon a time I let picasa rewrite the Exif data for a number of albums that I curated -- after indexing all of my photos, I now have two sets of all of my curated photos -- the ones for the original copies and the ones for the rewritten copies -- even though the image data is identical. Since the Exif metadata was updated inside of the file, the sha1 hashes are different and the files are no longer considered duplicates.
2. On a separate occasion I painstakingly "fixed" the orientation of a lot of my pictures that were taken in portrait mode (without an accelerometer). Once again, because this was just an update to the Exif orientation metadata, the image data is identical, but the file content, and thus the hash, is different. No duplicate detected. 

You bring up an interesting idea when it comes to using *date* information, but one of your feature requests (facial recognition combined with how old is) points out that you are very aware that date information is often questionable in origin. I think the key points you were trying to make were:

1. Older files are more likely to be higher quality original source files rather than re-encoded potentially resampled, lossy copies
2. Newer files are likely to have the benefit of manual curation and thus better metadata quality

The question is how to identify newer files from older ones -- I have a lot of pictures from older cameras with unreliable clocks that often reported some low date in the images metadata (like the cameras year of manufacturing) so the date is always earlier than the date of the picture. As a counter example, I have scanned a lot of old photos that predated modern computers, so the date of the picture is much earlier than any date available in the metadata unless I manually edited it.

**Reply by dscheffy** [reply]
I should have prefaced with the fact that I've only just started playing around with photoprism... I just reindexed with the "stack sequences" disabled, deleted the index database (but not the sidecar files or thumbnails) and reran the index step. I was surprised to see that while indexing it was once again stacking the sequences, but by this time around I noticed that an overwhelming number of stacked photos were ones with different file hashes, but identical image blob content (this did have the downside of rendering all of my rotated photos with the wrong primary and no ability to adjust, but I imagine that will be fixed over time).

**Reply by lastzero** [reply]
Fixed a ton of merging / stacking issues today, and pushed fresh Docker images for testing. You may use the `photoprism reset` command to reset your index. It makes sure sidecar YAML files get deleted. Alternatively, delete the sidecar folder and database contents manually. The YAML files contain an index backup, so that files might be stacked that were stacked before - independent of your new settings. We've introduced a new "single" flag so that the indexer knows, that no additional files should be stacked.

**Reply by alexisLefebvre** [reply]
Could you please describe what has been changed in sidecar files? Maybe we can edit the sidecar files to keep relevant data instead of resetting everything.

**Reply by lastzero** [reply]
We didn't change anything in the sidecar files. But they can influence stacking as they contain the photo UID. Files with the same photo UID will be merged together. But these YAML files will only be used as backup. It's not a way to update database contents after indexing for the first time.

**Reply by lastzero** [reply]
Refactored the indexer once again to get rid of timing issues when indexing on multiple CPUs in parallel. Demo looks good to me. Let's see if you find more issues. Note that files sharing the same base name (file name without extension), like `photo.raw` and `photo.jpg`, can't be unstacked or indexed separately upfront, without being renamed as we need to rely on common names to identify sidecar files. When clicking the unstack button in the app, one file gets an extended name automatically. It doesn't make sense to index sidecar files without a main file they belong to. I'm aware some users have files like `photo.jpg` and `photo.jpeg` in the same directory that are totally unrelated. In that case just rename them, even a human would not be able to see they are unrelated.

**Comment by lastzero** [comment]
Thanks for your feedback! I currently don't have time to reply in detail, but we'd love to implement some of this next year.
